go/routers/app_multiplexer/vumi_app.py

```
@@ -0,0 +1,261 @@
+            config.redis_manager,
+            key_prefix="application_multiplexer",
+            max_session_length=config.session_expiry
+        )
+        self.handlers = {
+            self.STATE_START: self.handle_state_start,
+            self.STATE_SELECT: self.handle_state_select,
+            self.STATE_SELECTED: self.handle_state_selected,
+            self.STATE_BAD_INPUT: self.handle_state_bad_input,
+        }
+
+    def target_endpoints(self, config):
+        """
+        Make sure the currently active endpoint is still valid.
+        """
+        return set([entry['endpoint'] for entry in config.entries])
+
+    @inlineCallbacks
+    def handle_inbound(self, config, msg, conn_name):
+        log.msg("Processing inbound message: %s" % (msg,))
+
+        user_id = msg['from_addr']
+        session = yield self.session_manager.load_session(user_id)
+        session_event = msg['session_event']
+        if not session or session_event == TransportUserMessage.SESSION_NEW:
+            log.msg("Creating session for user %s" % user_id)
+            session = {}
+            state = self.STATE_START
+            yield self.session_manager.create_session(user_id)
+        elif session_event == TransportUserMessage.SESSION_CLOSE:
+            yield self.handle_session_close(config, session, msg)
+            return
+        else:
+            log.msg("Loading session for user %s: %s" % (user_id, session,))
+            state = session['state']
+
+        try:
+            next_state, updated_session = yield self.handlers[state](
+                config, session, msg)
+            if next_state is None:
+                # Halt session immediately
+                log.msg(("Router configuration change forced session abort "
+                         "for user %s" % user_id))
+                yield self.session_manager.clear_session(user_id)
+                yield self.publish_error_reply(msg, config)
```

I don't think this is the right place for this logic. The handler should either raise an exception or it should log the error and send the response message itself.